bftengine/src/bcstatetransfer/RVBManager.cpp

```diff
@@ -0,0 +1,711 @@
+    std::istringstream rvb_data(std::string(reinterpret_cast<const char*>(desc.rvbData.data()), desc.rvbData.size()));
+    loaded_from_data_store = in_mem_rvt_->setSerializedRvbData(rvb_data);
+    if (!loaded_from_data_store) {
+      LOG_ERROR(logger_, "Failed to load RVB data from stored checkpoint" << KVLOG(desc.checkpointNum));
+    } else {
+      rvb_data_source_ = RvbDataInitialSource::FROM_STORAGE_CP;
+    }
+
+    if (!in_mem_rvt_->validate()) {
+      // in some cases, if not fetching, we might try to look for other checkpoints but this is fatal enough to
+      // reconstruct from storage
+      LOG_ERROR(logger_, "Failed to validate loaded RVB data from stored checkpoint" << KVLOG(desc.checkpointNum));
+      loaded_from_data_store = false;
+      in_mem_rvt_->clear();
+      rvb_data_source_ = RvbDataInitialSource::NIL;
+    }
+  }
+
+  if (!loaded_from_data_store && (desc.maxBlockId > 0)) {
+    // If desc data is valid, try to reconstruct by reading digests from storage (no persistency data was found)
+    LOG_WARN(logger_, "Reconstructing RVB data" << KVLOG(loaded_from_data_store, desc.maxBlockId));
+    addRvbDataOnBlockRange(
+        as_->getGenesisBlockNum(), desc.maxBlockId, std::optional<STDigest>(desc.digestOfMaxBlockId));
+    if (!in_mem_rvt_->validate()) {
+      LOG_FATAL(logger_, "Failed to validate reconstructed RVB data from storage" << KVLOG(desc.checkpointNum));
+    }
+    rvb_data_source_ = RvbDataInitialSource::FROM_STORAGE_RECONSTRUCTION;
+  }
+
+  LOG_INFO(logger_, std::boolalpha << KVLOG(pruned_blocks_digests_.size(), desc.checkpointNum, loaded_from_data_store));
+  if (print_rvt && (debug_prints_log_level.find(getLogLevel()) != debug_prints_log_level.end())) {
+    in_mem_rvt_->printToLog(false);
+  }
+}
+
+// Remove (Prune) blocks from RVT, and from pruned_blocks_digests_ data structure
+void RVBManager::pruneRvbDataDuringCheckpoint(const CheckpointDesc& new_checkpoint_desc) {
+  LOG_TRACE(logger_, "");
+  std::lock_guard<std::mutex> guard(pruned_blocks_digests_mutex_);
+  size_t i{};
+
+  if (pruned_blocks_digests_.empty()) {
+    return;
+  }
+  ConcordAssertGE(new_checkpoint_desc.checkpointNum, last_checkpoint_desc_.checkpointNum);
+  ConcordAssertGE(new_checkpoint_desc.maxBlockId, last_checkpoint_desc_.maxBlockId);
+
+  // First, Remove old blocks from pruned_blocks_digests_, then persist it. These blocks were already removed from RVB
+  // data and the removal is reflected in the RVB persisted tree of the last checkpoint.
+  //
+  // We must prune the pruning vector based on some persisted checkpoint in the past.
+  // We cannot allow losing digests, we will never get them back (blocks are not in storage anymore) and will have
+  // to reconstruct the whole tree.
+  if (last_checkpoint_desc_.checkpointNum > 0) {
+    for (i = 0; i < pruned_blocks_digests_.size(); ++i) {
+      if (pruned_blocks_digests_[i].first >= last_checkpoint_desc_.maxBlockId) {
```

Comment:
I assume you mean last stored checkpoint?
The answer is no.
We must assume that multiple components in the system are behaving as expected, else, nothing will work.
This is a major assumption, that It cant go beyond.
But even it will go beyond in the future - the code is ready :
I wrote my code in such way that they will decide to do so in the future, so we include in this checkpoint exactly the blocks between last_checkpoint_desc_.maxBlockId and new_checkpoint_desc.maxBlockId